lighthouse-core/gather/gatherers/console.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+/**
+ * @fileoverview Gathers all entries logged to the console, including console API calls,
+ * exceptions, and browser reports (on violations, interventions, deprecations, etc.).
+ */
+
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+
+/**
+ * @param {LH.Crdp.Runtime.RemoteObject} obj
+ * @return {string}
+ */
+function remoteObjectToString(obj) {
+  if (obj.value) {
+    // Primitive values will simply store the object.
+    return obj.value;
+  }
+  if (obj.type == 'function') {
+    return obj.description || '';
+  }
+  // Simulate calling String() on the object.
+  return `[${obj.type} ${obj.description}]`;
+}
+
+class Console extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Artifacts.ConsoleMessage[]} */
+    this._logEntries = [];
+
+    this._onConsoleAPICalled = this.onConsoleAPICalled.bind(this);
+    this._onExceptionThrown = this.onExceptionThrown.bind(this);
+    this._onLogEntryAdded = this.onLogEntry.bind(this);
+  }
+
+  /**
+   * Handles events for when a script invokes a console API.
+   * @param {LH.Crdp.Runtime.ConsoleAPICalledEvent} event
+   */
+  onConsoleAPICalled(event) {
+    const level = event.type;
+    if (level !== 'warning' && level !== 'error') {
+      // Only gather warnings and errors for brevity.
+      return;
+    }
+    /** @type {LH.Crdp.Runtime.RemoteObject[]} */
+    const args = event.args || [];
+    const text = args.map(remoteObjectToString).filter(Boolean).join(' ');
+    if (!text) {
+      return;
+    }
+    const {url, lineNumber, columnNumber} =
+      event.stackTrace && event.stackTrace.callFrames[0] || {};
+    /** @type {LH.Artifacts.ConsoleMessage} */
+    const consoleMessage = {
+      source: level === 'warning' ? 'console.warn' : 'console.error',
+      level,
+      text,
+      stackTrace: event.stackTrace,
+      timestamp: event.timestamp,
+      url,
+      lineNumber,
+      columnNumber,
+    };
+    this._logEntries.push(consoleMessage);
+  }
+
+  /**
+   * Handles exception thrown events.
+   * @param {LH.Crdp.Runtime.ExceptionThrownEvent} event
+   */
+  onExceptionThrown(event) {
+    const text = event.exceptionDetails.exception ?
+          event.exceptionDetails.exception.description : event.exceptionDetails.text;
+    if (!text) {
```

Comment:
do you think it's still worth returning if we still have a URL/stack trace?

Reply:
Good idea. Done